Monika After Story/game/connect4.rpy

```diff
@@ -0,0 +1,690 @@
+        # pygame events
+        MOUSE_EVENTS = (
+            pygame.MOUSEMOTION,
+            pygame.MOUSEBUTTONUP,
+            pygame.MOUSEBUTTONDOWN
+        )
+        
+        # variables to indicate Monika and player's pins/turn
+        AI = 1
+        PLAYER = 2
+        
+        # indicate the board's rows columns
+        CEL_COL = 7
+        CEL_ROW = 6
+        
+        # the  board's dimensions
+        BOARD_BORDER_WIDTH = 15
+        BOARD_BORDER_HEIGHT = 15
+        CEL_WIDTH = 60
+        CEL_HEIGHT = 60
+        BOARD_WIDTH = BOARD_BORDER_WIDTH * 2 + CEL_WIDTH * CEL_COL
+        BOARD_HEIGHT = BOARD_BORDER_HEIGHT * 2 + CEL_HEIGHT * CEL_ROW
+        
+        # offset for board's position
+        DISP_X_OFFSET = 200
+        DISP_Y_OFFSET = 200
+        
+        BOARD_X_POS = int(1280 - BOARD_WIDTH - DISP_X_OFFSET)
+        BOARD_Y_POS = int(720 - BOARD_HEIGHT - DISP_Y_OFFSET)
+        
+        # position of the upper left most cel
+        CEL_BASE_X_POS = BOARD_X_POS + BOARD_BORDER_WIDTH
+        CEL_BASE_Y_POS = BOARD_Y_POS + BOARD_BORDER_HEIGHT
+        
+        # dimensions of buttons the player use to drop pins on their turn (AKA col button)
+        COL_BUTTON_WIDTH = 60
+        COL_BUTTON_HEIGHT = 53
+        COL_BUTTON_X_SPACING = 5
+        COL_BUTTON_Y_SPACING = 7
+        COL_BUTTON_X_DISTANCE = int(COL_BUTTON_WIDTH + COL_BUTTON_X_SPACING)
+        
+        # menu button dimensions
+        BUTTON_WIDTH = 120
+        BUTTON_HEIGHT = 35
+        BUTTON_X_SPACING = 10
+        BUTTON_Y_SPACING = 10
+        BUTTON_Y_DISTANCE = int(BUTTON_HEIGHT + BUTTON_Y_SPACING)
+        
+        # col button position
+        COL_BUTTON_X = int(BOARD_X_POS + BOARD_WIDTH / 2 - COL_BUTTON_WIDTH / 2)
+        COL_BUTTON_Y = int(BOARD_Y_POS - COL_BUTTON_HEIGHT - COL_BUTTON_Y_SPACING)
+        
+        # menu button position
+        BUTTON_X = BOARD_X_POS + BOARD_WIDTH + BUTTON_X_SPACING
+        BUTTON_Y = BOARD_Y_POS + BOARD_HEIGHT - BUTTON_HEIGHT
+        
+        # images definition
+        AI_IMAGE = Image("mod_assets/games/connect4/monika_pin.png")
+        PLAYER_IMAGE = Image("mod_assets/games/connect4/player_pin.png")
+        BOARD_IMAGE = Image("mod_assets/games/connect4/connect4_board.png")
+        WIN_TILE_IMAGE = Image("mod_assets/games/connect4/win_tile.png")
+        COL_BUTTON_IDLE_IMAGE = Image("mod_assets/games/connect4/arrow_down_idle.png")
+        COL_BUTTON_HOVER_IMAGE = Image("mod_assets/games/connect4/arrow_down_hover.png")
+        COL_BUTTON_INSENSITIVE_IMAGE = Image("mod_assets/games/connect4/arrow_down_insensitive.png")
+        
+        def __init__(self, first_turn):
+            # prepare the Displayable
+            super(MASConnect4Displayable, self).__init__()
+            
+            # init the variables
+            self.is_game_over = False
+            self.quit_game = False
+            self.turn = first_turn
+            self.board = dict()
```

Comment:
why is the board a dictionary? a connect 4 board makes more sense as a 2d array, or even just a 1d array with 2d indexing

if this is a 2d array. then we can do much better row/col/diag checks since we only have to deal with x, y coordinates and not the weird "line up" functions.

Reply:
It _was_ a 2d array, but then [multi suggested to id the board using a dict.](https://github.com/Monika-After-Story/MonikaModDev/pull/6929#issuecomment-769311409) I think that was a good suggestion. Using a 2d array earlier was a whole lot of mess to iterate through.

Reply:
This should be reverted back. A 2d array is understandable and easy to map to an actual board. A dictionary makes sense as an index, but when the most number of spots we have in this board is 42, the performance difference between a dictionary and a 2d array is negligible. 

And actually I've written C code as an assignment to play connect 4. It even uses a 1d array to represent the board, but regardless, it's not that complicated to do.

Iterate over rows/cols in a 2d array? Very easy - your code didn't do piece counting or size match checks but we can do that to make it even more efficient:
```c
int check_columns(char player_id){
    /*
     * For the columns, we only need to check from the bottom up until 
     * CHECKSIZE-1 from the top
     */

    /* controls */
    int col;    /* current column */
    int row;    /* current row */

    /* conditions */
    int has_more_pieces;    /* 1 when there are more pieces int eh column,
                               0 if not */

    /* containers */
    int matching;   /* how many matches we found */

    /* for every column */
    for(col = 1; col <= COLSIZE; col++){

        /* first check if anything is even in here */
        if(board[(ROWSIZE-1)*COLSIZE + col] != EMPTY){
            matching = 0;
            has_more_pieces = 1;
            row = ROWSIZE-1;

            /* while more pieces in the column */
            while(has_more_pieces && row >= 0){
//                printf("tset %i\n",matching);
                /* check the element, counting for CHECKSIZE match */
                if(board[row*COLSIZE + col] == EMPTY 
                        || (row < CHECKSIZE-1 && matching < 1)){
                    /* no more pieces or no more room for matches */
                    has_more_pieces = 0;
                }else if(board[row*COLSIZE + col] == player_id){
                    /* match found */
                    matching++;
                }else{
                    /* no match found */
                    matching = 0;
                }

                if(matching == CHECKSIZE){
                    return 1;
                }

                row--;
            } /* while more pieces in the column*/
        } /* check if anything is in the column */
    } /* for every column */

    return 0;
} /* check_columns */
```

Diagnols can be done with 4 separate loops in a similar way. The loops check:  left column negative slope, left column positive slope, first row neg slope, first row pos slope. This is an example of a loop for left col neg:
```c
    /* first, do negative slope scans */
    /* left column scans */
    /* for start spaces on left column */
    for(start_row = ROWSIZE-CHECKSIZE; start_row >= 0; start_row--){
        has_space = 1;
        matching = 0;
        row = start_row;
        col = 0;

        /* while more space in the diag */
        while(has_space && row < ROWSIZE){
//            printf("test %i\n", matching);

            /* check the element, counting for CHECKSIZE match */
            if(ROWSIZE-row < CHECKSIZE && matching < 1){
                /* no more room for matches */
                has_space = 0;
            }else if(board[row*COLSIZE + col] == player_id){
                /* match found */
                matching++;
            }else{
                /* no match found */
                matching = 0;
            }

            if(matching == CHECKSIZE){
                return 1;
            }

            row++;
            col++;
        } /* while more space in the diag */
    } /* for start spaces on left column */
```
Since this is my code, I could just rewrite it in python and we can use it here. Theres a few simplifications that can be made, too, like using generic bounds checking. The gist of all of these is to count consecutive pieces while avoiding unnecessary iterations by not checking areas smaller than the CHECKSIZE. CHECKSIZE being the number of pieces in a row to consider a win, which means this is also generic to any version of connect number. 

Additionally, if you just want xy tuples of diagnols for examination in a much better way: might be useful for the AI part, definitely **not** good to use with win checking since it would add a third loop.
```python
def generate_diag(self, start_x, start_y, positive_slope):
    # also assuming bottom left is (0,0)
    if not self.in_bounds(start_x, start_y):
        return []
    
    if positive_slope:
        slope_mod = 1
    else:
        slope_mod = -1

    x = start_x
    y = start_y
    xy_tups = []
    while True:
        x += 1
        y += slope_mod

        if self.in_bounds(x, y):
            xy_tups.append((x, y))
        else:
            return xy_tups

    return xy_tups
```

Reply:
I hope I'm not misunderstanding what your code is doing, but I see that it only returns whether or not there is a match the size of `CHECKSIZE` and doesn't include which cells are those matches found in, while also doing tile checking from the start, even though getting the tiles from another function is much more reusable (which I didn't do in the earlier version that still uses 2d array).

In the current version I've worked on, four functions are used to get the tiles, which are then fed into both `check_connect` (a function used to check if there's a match) and `get_winning_line_up` (used to check if there's gonna be a win in one more move, used by the AI)
```
        # GET LINE UP FUNCTIONS START
        # These functions return a LIST of LINE UPS
        # a LINE UP is a LIST of the BOARD'S (KEY, VALUE) in tuple pair
        # didn't use dict() because it's unordered and the order needs to be preserved
        
        def get_vertical(self):
            """
            Returns vertical line ups
            """
            board_form = [[(self.g_k(j, i), self.board.get(self.g_k(j, i))) for j in range(self.CEL_ROW)] for i in range(self.CEL_COL)]
            return board_form
                    
        def get_horizontal(self):
            """
            Returns horizontal line ups
            """
            board_form = [[(self.g_k(i, j), self.board.get(self.g_k(i, j))) for j in range(self.CEL_COL)] for i in range(self.CEL_ROW)]
            return board_form
                
        def get_diagonal_ltr(self):
            """
            Returns diagonal top Left To bottom Right (ltr) line ups
            """
            board_form = list()
            for i in range(4):
                line_up = list()
                line_up2 = list()
                i2 = 0
                for j in range(6):
                    line_up.append((self.g_k(i+i2, j), self.board.get(self.g_k(i+i2, j))))
                    line_up2.append((self.g_k(j, i+i2), self.board.get(self.g_k(j, i+i2))))
                    i2 += 1
                if i != 0:
                    board_form.append(line_up)
                board_form.append(line_up2)
            return board_form
            
        def get_diagonal_rtl(self):
            """
            Returns diagonal top Right To bottom Left (rtl) line ups
            """
            board_form = list()
            for i in range(6):
                line_up = list()
                i2 = list(range(3, 9))
                for j in range(9):
                    line_up.append((self.g_k(i2[i]-j, j), self.board.get(self.g_k(i2[i]-j, j))))
                board_form.append(line_up)
            return board_form
                
        # GET LINE UP FUNCTIONS END
```
the `generate_diag` function you suggested, on the other hand, I could see that the current diagonal functions can use it. I admit my method was quite hacky and I have been having difficulty finding a clean solution for that part.

Reply:
Since we're using python, this part can easily be changed to return the position as well:
```c
            if(matching == CHECKSIZE){
                return 1;
            }
```

Regardless, this doesn't change the fact that a 2d backing array should be the primary board representation. Getting xy tuples for columns/rows is just as trivial as it is for diagonals. But we should use boop's suggestion too as it makes both win checking and getting potential winning moves extremely trivial.

Reply:
Ahh, that is correct. I see it now.

So, regarding the reversion back to using 2d array, I changed it to dict from @multimokia's suggestion and I actually quite like it. In any case, the two of you have conflicting suggestions... I think it would be better to discuss this with @multimokia as well

Reply:
I...think I have a pretty clear grasp on it. I mean, I did make it 2d in the first place. Just gonna have to redo the optimization I've worked on the dict version back into the 2d method. Might take a while tho.

So, are we still gonna do it? Just asking for your official green light before I commit myself to work on it.

Reply:
2D array for sure. Boop's suggestion is optional.

Reply:
Alright, let me see what I can do.